### DISCUSSION TITLE
Best practice for dealing with packages re-exporting large number of modules

### DISCUSSION THREAD

**Discussion by jdb8** [webpack/webpack#14285] [Q&A]
Hi! Hopefully the title makes sense: I'm hoping someone will have encountered a similar scenario and have some suggestions.

I'm trying to improve the performance of my webpack builds (both production and development) and have been digging into why builds are taking longer than expected. I've come to the conclusion that the biggest issue in my codebase (which is a monorepo) is the existence of certain packages which re-export many submodules from their main `index.js` entrypoint.

For example, there are many packages which consist of a folder structure like this:

```
- my-package/
    - componentA/
    - componentB/
    - componentC/
    - index.js
```

where `index.js` consists of:

```js
import ComponentA from './componentA';
import ComponentB from './componentB';
import ComponentC from './componentC';

export { ComponentA, ComponentB, ComponentC };
```

Across the codebase, various other packages import `Component{A,B,C}` via `import { ComponentA } from 'my-package';`.

This works fine, and crucially with the addition of `sideEffects: false` in the right places, also tree-shakes correctly in production: none of the code for components that are not used is included in the final prod bundle.

However, the issue I'm realising (which should have been obvious, I suppose) is that the existence of the re-exports in the main `index.js` still forces webpack to traverse and parse each of the other components, even when none of them are used. In a large monorepo, the cost of parsing modules adds up and contributes to slow builds, even in development.

The following would "fix" it:

```js
// calling site
import ComponentA from 'my-package/componentA';
```

Now `ComponentB` and `ComponentC` won't even be parsed by Webpack. However, this fix isn't ideal:

* I'd ideally like to avoid having to manually update all call sites and introduce lint rules etc. to guarantee they remain as deep imports
* It's harder to enforce the api of the component without adding package exports, because any subfile can be imported

One strategy I used for a particularly egregious example of this was to apply the [babel-plugin-transform-imports](https://www.npmjs.com/package/babel-plugin-transform-imports) plugin, and have just found [babel-plugin-direct-import](https://www.npmjs.com/package/babel-plugin-direct-import) which I'm yet to try, but I'm wondering if there's a better way to approach this that doesn't require rewriting the module source at all, and doesn't require using Babel or any specific transpiler plugin.

So: is there any way to make Webpack skip parsing the modules that aren't being imported in this case? Right now I feel like the only options are:

* Manually update call sites to use deep imports and skip the re-exporting `index.js` entirely from parsing
* Find a way to use `module.noParse` on the irrelevant files? But unclear if this would work for most real scenarios
* Use a transpiling loader to update the calling import sites to deep import behind the scenes
* Write a webpack custom resolver plugin to do deep importing similarly (does one exist in the community?)

It definitely seems like a common reason for build perf given that best practice for a while appears to have been to re-export heavily (since tree shaking will optimise things away) - what I haven't seen talked about is the cost you still pay at build time. Curious if anyone else has run into this and how they've solved it if so.

**Comment by alexander-akait** [comment]
Hm, it should not decrease perf very, maybe you can provide small example?

**Reply by jdb8** [reply]
Yeah, it's not really that webpack is doing anything wrong or necessarily inefficient, but in a case where you have some packages consisting of hundreds or thousands of exported modules doing any kind of parsing on them adds up. It results in a node perf flamegraph consisting of thousands of tiny slivers of parse time.

The closest example I can give is to create a new project with something like `lodash-es` as a dependency, and import one function: if you add a console.log to [this line](https://github.com/webpack/webpack/blob/69777d58f707f53b861c0464b3472f4de4d15b67/lib/NormalModule.js#L1037) it shows that every lodash function file is being parsed unnecessarily:

```js
// Import one named export, but since this comes from https://unpkg.com/browse/lodash-es@4.17.21/lodash.js
// it requires webpack to parse the entire file and all other lodash functions
import {flatMap} from 'lodash-es';
```

```bash
❯ webpack
parsing /Users/jbateson/code/demo-large-number-of-modules/src/index.js
parsing /Users/jbateson/code/demo-large-number-of-modules/node_modules/lodash-es/lodash.js
parsing /Users/jbateson/code/demo-large-number-of-modules/node_modules/lodash-es/after.js
parsing /Users/jbateson/code/demo-large-number-of-modules/node_modules/lodash-es/ary.js
parsing /Users/jbateson/code/demo-large-number-of-modules/node_modules/lodash-es/assignIn.js
parsing /Users/jbateson/code/demo-large-number-of-modules/node_modules/lodash-es/assign.js
parsing /Users/jbateson/code/demo-large-number-of-modules/node_modules/lodash-es/at.js
...
```

This is maybe not the best example since each individual lodash file is small, and doesn't itself import other modules, but hopefully you can see that in a larger project with more interdependencies you can end up with this taking up some chunk of time.

**Reply by alexander-akait** [reply]
What is webpack version? If each file processed, it means you have import them, no magic here
